contracts/Staking.sol

```diff
@@ -0,0 +1,507 @@
+        require(isContract(_stakingToken));
+        stakingToken = _stakingToken;
+    }
+
+    /* External functions */
+
+    /**
+     * @notice Stakes `_amount` tokens, transferring them from `msg.sender`
+     * @param _amount Number of tokens staked
+     * @param _data Used in Staked event, to add signalling information in more complex staking applications
+     */
+    function stake(uint256 _amount, bytes _data) external {
+        _stakeFor(msg.sender, _amount, _data);
+    }
+
+    /**
+     * @notice Stakes `_amount` tokens, transferring them from caller, and assigns them to `_account`
+     * @param _account The final staker of the tokens
+     * @param _amount Number of tokens staked
+     * @param _data Used in Staked event, to add signalling information in more complex staking applications
+     */
+    function stakeFor(address _account, uint256 _amount, bytes _data) external {
+        _stakeFor(_account, _amount, _data);
+    }
+
+    /**
+     * @notice Unstakes `_amount` tokens, returning them to the user
+     * @param _amount Number of tokens staked
+     * @param _data Used in Unstaked event, to add signalling information in more complex staking applications
+     */
+    function unstake(uint256 _amount, bytes _data) external {
+        // unstaking 0 tokens is not allowed
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // transfer tokens
+        // it will fail for non ERC20 compliant tokens which don't return anything
+        // see: https://medium.com/coinmonks/missing-return-value-bug-at-least-130-tokens-affected-d67bf08521ca
+        require(stakingToken.transfer(msg.sender, _amount), ERROR_TOKEN_TRANSFER);
+
+        // checkpoint updated staking balance
+        _modifyStakeBalance(msg.sender, _amount, false);
+
+        // checkpoint total supply
+        _modifyTotalStaked(_amount, false);
+
+        emit Unstaked(msg.sender, _amount, totalStakedFor(msg.sender), _data);
+    }
+
+    /**
+     * @notice Lock `_amount` staked tokens and assign `_manager` as manager with `_data` as data, so they can not be unstaked
+     * @param _amount The amount of tokens to be locked
+     * @param _manager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
+     * @param _data Data to parametrize logic for the lock to be enforced by the manager
+     * @return The id of the newly created lock
+     */
+    function lock(uint256 _amount, address _manager, bytes _data) external returns (uint256) {
+        // locking 0 tokens is invalid
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // check enough unlocked tokens are available
+        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
+
+        accounts[msg.sender].lastLockId++;
+        uint256 _lockId = accounts[msg.sender].lastLockId;
+        accounts[msg.sender].activeLockIds.push(_lockId);
```

Comment:
We should cap the amount of locks allowed per account to avoid OOG errors.

Reply:
Good point! I'll do some math to estimate what the cap should be.
Another idea I was considering, in case we finally drop the `unlockAll*` functions, as I think it would only impact on `unlockedBalanceOf`, was to find an alternative like keeping the unlocked balance up to date, but I guess it would be a waste of gas. Let's see what this number is.